### PR TITLE
Changes to NpuDmaMemcpyNdOp and AIEDmaToNpu to support sub-word strides, offsets and sizes

### PR DESCRIPTION
**Pull request by pvasireddy-amd**
This PR removes the manual 32-bit alignment of `strides`, `offsets` and `sizes` for `npu_dma_memcpy_nd`. `AIEDmaToNpu.cpp` now automatically aligns `strides`, `offsets[0]` and `sizes[0]` based on the provided element width. 

The previous condition requiring a fixed element bit width of 32 in `AIEXDialect.cpp`  is replaced with two checks: 1. Minimum data transfer size (number of elements * element width) must be at least 32. 2. Element width remains within 32.

Removed hard-coded 32-bit address granularity  assumption. Added `getAddressGenGranularity()` to determine the appropriate alignment based on the target device.

Manual conversions of `strides`, `offsets` and `sizes` has been removed from the `mlir-aie/programming_examples` directory.

New test cases are added to verify sub-word `strides`, `offsets` and `sizes` in `NpuDmaMemcpyNdOp` and the warning for `objectFifoLink` and updated the existing tests.

### CODE REVIEW

File: include/aie-c/TargetModel.h [hunk 1]
```diff
@@ -42,6 +42,9 @@ DEFINE_C_API_STRUCT(AieTargetModel, uint64_t);
 
 MLIR_CAPI_EXPORTED AieTargetModel aieGetTargetModel(uint32_t device);
 
+/// Returns the data bus width for the target model.
+MLIR_CAPI_EXPORTED uint32_t aieGetTargetModelDataBusWidth(AieTargetModel targetModel);
```
Comment (jgmelber) [inline include/aie-c/TargetModel.h:46]: This isn't exactly related to a data bus or stream width, it is the address generation granularity for the DMAs. Perhaps this should be ```aieGetTargetModelAddressGenGranularity```, or something

File: include/aie/Dialect/AIE/IR/AIETargetModel.h [hunk 2]
```diff
@@ -61,6 +61,9 @@ class AIETargetModel {
   /// Return the target architecture.
   virtual AIEArch getTargetArch() const = 0;
 
+  /// Return the data bus width of the device.
+  virtual uint32_t getDataBusWidth() const = 0;
```
Comment (jgmelber) [inline include/aie/Dialect/AIE/IR/AIETargetModel.h:65]: Address generation granularity

Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIETargetModel.h]: umm `getMinStrideBitWidth`?

**Comment by makslevental** [comment]
So for which arch is the minimum stride not 32? Some forthcoming one that's not public yet?

File: include/aie-c/TargetModel.h (hunk 1 above)
Comment (pvasireddy-amd) [reply include/aie-c/TargetModel.h]: Changed

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp [hunk 3]
```diff
@@ -1194,6 +1194,27 @@ struct AIEObjectFifoStatefulTransformPass
...
 
+    //===------------------------------------------------------------------------===//
+    // Check whether the same data type is used in I/O FIFOs of the objectFifoLink
+    //===------------------------------------------------------------------------===//
+    // We compare the data type of the input and the output FIFOs of the 
+    // objectFifoLink and emits a warning if there is a mismatch
+
+    for(auto linkOp : device.getOps<ObjectFifoLinkOp>()){
+      auto fifoIn = linkOp.getInputObjectFifos()[0];
+      auto fifoOut = linkOp.getOutputObjectFifos()[0];
```
Comment (AndraBisca) [inline lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:1205]: There may be multiple fifoIns and fifoOuts  (not at the same time), so for completeness you should check them all.

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp [hunk 4]
```diff
@@ -1194,6 +1194,27 @@ struct AIEObjectFifoStatefulTransformPass
     std::set<TileOp>
         objectFifoTiles; // track cores to check for loops during unrolling
 
+    //===------------------------------------------------------------------------===//
+    // Check whether the same data type is used in I/O FIFOs of the objectFifoLink
+    //===------------------------------------------------------------------------===//
+    // We compare the data type of the input and the output FIFOs of the 
+    // objectFifoLink and emits a warning if there is a mismatch
```
Comment (AndraBisca) [inline lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:1201]: Minor nitpick: `emit` without the 's'

**Review by AndraBisca — APPROVED** [review]
Looks good apart from the two small comments I made, thank you for updating all the tests!

**Comment by jgmelber** [comment]
RTM?

**Comment by pvasireddy-amd** [comment]
Now it is ready to merge. Updated tests since no-prolog branch is now merged into main

**Comment by andrej** [comment]
Is it possible that **offset** is in bytes?

[Multiplying my offsets by two](https://github.com/Xilinx/mlir-aie/pull/1551/commits/f678f39041cb254df2e0818779988f0dd244fe59) is the only way I seem to be able to get my matrix multiplication code to verify with different inputs.

This probably slid by because the verification threshold may be too high currently (15%). But if I give it different inputs, with tighter verification tolerance, it only passes if I multiply offsets by two (size of bf16). I will reduce the verification tolerance in my PR so errors like this don't slide by testing in the future.

I did see [this change](https://github.com/pvasireddy-amd/mlir-aie/blob/6f757afd15ada130f43c1f78070e46a0aa0aeda7/lib/Dialect/AIEX/Transforms/AIEDmaToNpu.cpp#L246), but somehow multiplying my offsets by two is the only way I can fix this. Maybe it should be the last element of the offsets array that should be modified?

**Pull request merged**